### PR TITLE
Less raw addresses in Numba Expression

### PR DESCRIPTION
**Pull request by michalhabera**

### CODE REVIEW

File: python/dolfin/function/expression.py [hunk 1]
```diff
@@ -75,23 +76,23 @@ def eval(values, x, cell_idx, num_points, value_size, gdim, num_cells):
         return decorator
 
 
-class Expression(cpp.function.Expression):
+class Expression:
     def __init__(self,
-                 eval_func: int,
+                 eval_func: numba.ccallback.CFunc = None,
```
Comment (garth-wells) [inline python/dolfin/function/expression.py]: Why `None` by default?

File: python/dolfin/function/expression.py [hunk 2]
```diff
@@ -112,7 +113,21 @@ def __init__(self,
...
+        if eval_func is not None:
+            self._eval_address = eval_func.address
+            self._cpp_object.set_eval(self._eval_address)
+
+    @property
+    def eval_address(self): # noqa
+        return self._eval_address
+
+    @eval_address.setter
+    def eval_address(self, value: int):
```
Comment (garth-wells) [inline python/dolfin/function/expression.py]: Can we do everything via the initialiser and drop this?

File: python/dolfin/function/expression.py [hunk 3]
```diff
@@ -7,6 +7,7 @@
 from typing import Callable
 
 import numba
+import numba.ccallback
```
Comment (garth-wells) [inline python/dolfin/function/expression.py]: Is this needed? Fully qualified namespace is used below.

File: python/dolfin/function/expression.py (hunk 2 above)
Comment (michalhabera) [reply python/dolfin/function/expression.py]: I wanted to avoid isinstance check in __init__

Comment (garth-wells) [reply python/dolfin/function/expression.py]: Use `try`?

File: python/dolfin/function/expression.py (hunk 3 above)
Comment (michalhabera) [reply python/dolfin/function/expression.py]: Seems needed, numba.ccallback doesn't get imported with numba.

File: python/dolfin/function/expression.py [hunk 4]
```diff
@@ -112,4 +114,8 @@ def __init__(self,
         shape: tuple
             Value shape.
         """
-        super().__init__(eval_func, shape)
+        self.shape = shape
+        self._eval_func = eval_func
+
+        eval_address = getattr(eval_func, "address", eval_func)
```
Comment (garth-wells) [inline python/dolfin/function/expression.py]: Is it necessary to use `getattr`?

File: python/dolfin/function/expression.py [hunk 5]
```diff
@@ -114,20 +114,8 @@ def __init__(self,
         shape: tuple
             Value shape.
         """
-        self._cpp_object = cpp.function.Expression(shape)
-
         self.shape = shape
```
[the remark this answers is not in the document]
Comment (michalhabera) [reply python/dolfin/function/expression.py]: I think not needed at all

File: python/dolfin/function/expression.py (hunk 4 above)
Comment (michalhabera) [reply python/dolfin/function/expression.py]: If supplied `eval_func` has `eval_func.address` then it comes from JIT, otherwise take it as address itself. Its just short alternative to try catch.

Comment (garth-wells) [reply python/dolfin/function/expression.py]: It's not very clear to me. Better longer and clearer than short and hard to follow.

Comment (michalhabera) [reply python/dolfin/function/expression.py]: Ok, thats true.

File: python/dolfin/function/expression.py [hunk 6]
```diff
@@ -112,4 +114,11 @@ def __init__(self,
         shape: tuple
             Value shape.
         """
-        super().__init__(eval_func, shape)
+        self._eval_func = eval_func
```
Comment (garth-wells) [inline python/dolfin/function/expression.py]: Add comment 'Hold reference to eval function to avoid premature garbage collection'.

File: python/dolfin/function/expression.py [hunk 7]
```diff
@@ -112,4 +114,11 @@ def __init__(self,
...
+
+        try:
```
Comment (garth-wells) [inline python/dolfin/function/expression.py]:
```python
try:
    self._cpp_object = cpp.function.Expression(eval_func.address, shape)
except: FoorError:
    self._cpp_object = cpp.function.Expression(eval_func, shape)
```

**Comment by garth-wells** [comment]
Merging now. Can polish more later is needed.
